File: apps/agni-server/src/core/domains/entities/notification.ts
import Entity, { TrackableProperty } from "./entity";

export default class Notification extends Entity {
    private title: string
    private content: string
    private isRead: TrackableProperty<boolean>
    private dateTime: Date

    constructor(id: string, title: string, content: string, isRead: boolean, dateTime: Date) {
        super(id)
        this.title = title
        this.content = content
        this.isRead = new TrackableProperty(isRead, this.markHasChange.bind(this))
        this.dateTime = dateTime
    }

    getTitle(): string {
        return this.title
    }

    getContent(): string {
        return this.content
    } 

    getIsRead(): boolean {
        return this.isRead.get() 
    }

    setIsRead(isRead: boolean) {
        this.isRead.set(isRead)
    }

    getDateTime(): Date {
        return this.dateTime
    }
}

File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/infras/persistences/jbdc_model/Notification.kt
package dev.auguste.agni_api.infras.persistences.jbdc_model

import dev.auguste.agni_api.core.entities.Notification
import dev.auguste.agni_api.infras.persistences.IMapper
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.util.UUID

@Table("notifications")
data class JdbcNotificationModel(
    @Id
    @get:JvmName("getIdentifier")
    @Column("notification_id")
    val id: UUID,

    @Column("title")
    val name: String,

    val content: String,

    @Column("is_read")
    val isRead: Boolean,

    val date: LocalDateTime
) : JdbcModel() {
    override fun getId(): UUID {
        return id
    }
}

@Component
class JdbcNotificationModelMapper: IMapper<JdbcNotificationModel, Notification> {
    override fun toDomain(model: JdbcNotificationModel): Notification {
        return Notification(
            id = model.id,
            title = model.name,
            content = model.content,
            dateTime = model.date,
            isRead =model.isRead,
        )
    }

    override fun toModel(entity: Notification): JdbcNotificationModel {
        return JdbcNotificationModel(
            id = entity.id,
            name = entity.title,
            content = entity.content,
            isRead = entity.isRead,
            date = entity.dateTime
        )
    }

    override fun getSortField(): Set<String> {
        return setOf("date")
    }
}